### ISSUE TITLE
Mercurial backend problems

### ISSUE THREAD

**Issue by andrewshadura** [nijel/weblate#684]
Current implementation of rebase doesn't handle manual merges by upstream, especially with Mercurial.

Here's the situation we have with [Kallithea translation](https://hosted.weblate.org/projects/kallithea/kallithea/).
![2015-03-27-103415_621x1022_scrot_edit](https://cloud.githubusercontent.com/assets/309253/6865549/1e2096f6-d46f-11e4-8513-b7f98eba168b.png)

Users contributed some translations which were later manually rebased and redacted by the upstream maintainer (changesets with sequential numbers 4912:4920 as generated by Weblate, 4903:4911 as committed by the maintainer). Later, some more changes were committed by the upstream (4923:4953) as well as by a Weblate user (4921:4922). In this situation, Weblate can't detect most of the changes it has locally have already been applied upstream, nor can it rebase them properly because or the merge conflict, apparently.

Ideally, I'd want Weblate to detect and remove the commits (4912:4920) already applied upstream, and then rebase the rest (4921:4922) to the tip of the default branch, possibly collapsing them together if they come from the same user to the same translation of the same language (see dotted lines leading to 4954).

**Comment by nijel** [comment]
Well with Git, it does detect that the changesets are already there and discards them on rebase. Unfortunately Mercurial is not that clever...

**Comment by andrewshadura** [comment]
It usually does, probably it needs some switch or something…

**Comment by nijel** [comment]
Webate does no magic here, it just calls ``hg rebase``.

**Comment by andrewshadura** [comment]
Okay, I see, those changes can't be automatically detected as some changesets were reordered and collapsed manually. Still it is possible to detect that, but it has to be done manually. At least a button ‘ignore changes from that to this one, they're already merged’ would do.

**Comment by nijel** [comment]
There is "Reset" button which does this :-).

**Comment by andrewshadura** [comment]
Where? Admin interface? I've got no access to it :)

**Comment by nijel** [comment]
Now you have :-) - https://hosted.weblate.org/projects/kallithea/#repository

**Comment by andrewshadura** [comment]
It doesn't help :(

**Comment by nijel** [comment]
How the reset does not help? It does not work?

**Comment by andrewshadura** [comment]
It does nothing.

**Comment by andrewshadura** [comment]
I think I'll retitle the issue back, as this issue isn't specific to Mercurial, I imagine the same scenario with Git, it's enough for commits to be collapsed, reordered or to have slight changes applied, even if the final result is the same.

**Comment by nijel** [comment]
The git does recognize if the same changes were applied, it does not matter if the commits are same. So sorry, but this is Mercurial specific issue or at least Git can out of the box handle most of the situations, while Mercurial can not.

As for reset, I've just realized that the way it is implemented for Mercurial is indeed wrong, do you know way how to reset local mercurial clone to match remote one discarding any local changesets done?

**Comment by nijel** [comment]
There reset for mercurial should now be fixed in git.

**Comment by andrewshadura** [comment]
I think I'll try to play with this when I have some free time. I'm pretty
sure I should be able to reproduce this issue with Git.

-- 
Cheers,
  Andrew

**Comment by andrewshadura** [comment]
As I mentioned, the changes aren't the same, the commits have been amended upstream, neither Git nor Mercurial can detect that.

**Comment by nijel** [comment]
Git detects when the change it is trying to apply on rebase is already there, but AFAIK Mercurial is not doing that, so you just end up with failure in this case. Git cares about the result, while Mercurial about the individual patches.

**Comment by nijel** [comment]
...but definitely there are situations where both would fail, but I don't think Weblate should handle those (and actually I have no clue how). That's why there is access to the underlying repository so that somebody can do manual merge (or rebase and reset on Weblate side).

**Comment by andrewshadura** [comment]
Have yo updated the hosted instance? The old commits are still hanging around, and I've lost my admin rights :(

Actually, I think, for Mercurial the way to solve the issue would be to use changeset obsoletion. When I rebase/collapse/amend commits, Mercurial marks old commits as obsoleted by the new commits and hides them. This feature isn't enabled by default as it's not yet stable enough internally, but it's already usable, so Weblate could use it, I guess.

**Comment by nijel** [comment]
It's not yet deployed on hosted Weblate, I'll probably push that tomorrow.

As for the admin rights, you have two accounts there, only one has admin rights...

**Comment by andrewshadura** [comment]
> Git detects when the change it is trying to apply on rebase is already there, but AFAIK Mercurial is not doing that, so you just end up with failure in this case. Git cares about the result, while Mercurial about the individual patches.

I have actually just tried. I exported the repository to Git and did a rebase. It failed, at the very same commit.

    $ git branch 
      bad
      good
    $ cat .git/refs/heads/bad
    8503d11333486e5d44b149aab10e317db6958e66
    $ cat .git/refs/heads/good
    b7b57a2d70400191d954eaf193a66c13830dae90
    $ git log bad | head -n 20
    commit 8503d11333486e5d44b149aab10e317db6958e66
    Author: Andrew Shadura <andrew@shadura.me>
    Date:   Fri Mar 13 14:24:32 2015 +0000

        i18n: updated translation for Slovak
        
        Currently translated at 3.8% (41 of 1077 strings)

    commit 372f6d2ecef62d7f84670575082fd15d4e7a4184
    Author: Andrew Shadura <andrew@shadura.me>
    Date:   Fri Mar 13 14:03:08 2015 +0000

        i18n: updated translation for Slovak
        
        Currently translated at 0.1% (2 of 1077 strings)

    commit 09800b26cb325fe1bbe7b64c7e44bafb280cc6ca
    Author: Andrew Shadura <andrew@shadura.me>
    Date:   Wed Mar 11 19:00:36 2015 +0000

    $ git log good | head -n 20
    commit b7b57a2d70400191d954eaf193a66c13830dae90
    Author: Mads Kiilerich <madski@unity3d.com>
    Date:   Wed Mar 25 20:47:45 2015 +0100

        setup: update dulwich requirement to >= 0.9.9 which fixes CVE-2015-0838

    commit 20d1227520d06eca961e63940cb49cd2c5936ca7
    Author: Mads Kiilerich <madski@unity3d.com>
    Date:   Wed Mar 25 20:38:09 2015 +0100

        controllers: consistently use formfill.render with force_defaults=False
        
        The inconsistency could cause confusion for developers. It seems to me like
        force_defaults=False should be the default ... and apparently it was that in
        older versions of formfill.
        
        It could perhaps make sense for us to have a wrapper that added the defualt
        values once, instead of repeating it all over ;-)

    commit 3586674429441cd86a6ec35db3b791069a6f5acb
    $ git checkout bad
    Checking out files: 100% (730/730), done.
    Switched to branch 'bad'
    $ git rebase good bad
    First, rewinding head to replay your work on top of it...
    Applying: Translated using Weblate (Russian)
    Using index info to reconstruct a base tree...
    M	kallithea/i18n/ru/LC_MESSAGES/kallithea.po
    Falling back to patching base and 3-way merge...
    Auto-merging kallithea/i18n/ru/LC_MESSAGES/kallithea.po
    CONFLICT (content): Merge conflict in kallithea/i18n/ru/LC_MESSAGES/kallithea.po
    Failed to merge in the changes.
    Patch failed at 0001 Translated using Weblate (Russian)
    The copy of the patch that failed is found in:
       /tmp/sdfs/.git/rebase-apply/patch

    When you have resolved this problem, run "git rebase --continue".
    If you prefer to skip this patch, run "git rebase --skip" instead.
    To check out the original branch and stop rebasing, run "git rebase --abort".

    $

**Comment by andrewshadura** [comment]
So no, Git is not smarter in this particular situation.

Regarding the accounts, could you please merge them?

**Comment by nijel** [comment]
@andrewshadura I've merged your accounts and deployed new version which fixes reset for mercurial.

As for rebase, I really don't see what Weblate should to in this case. I'm not going to provide some intefrace to inspect and select individual commits, so such cases (same as merge conflicts) have to be solved on vcs side, that's why the repository is exposed and the maintainers have controls to reset the repository over web interface.

**Comment by andrewshadura** [comment]
Well, reset now fails with:
> Failed to reset to remote branch on Kallithea/Kallithea.

And rebase causes this:
> Server Error
> The server had serious problems while serving your request. We've just sent our trained monkeys to fix the issue.

**Comment by nijel** [comment]
Dne Wed, 01 Apr 2015 01:36:52 -0700
Andrew Shadura <notifications@github.com> napsal(a):

> Well, reset now fails with:
> > Failed to reset to remote branch on Kallithea/Kallithea.
> 
> And rebase causes this:
> > Server Error
> > The server had serious problems while serving your request. We've just sent our trained monkeys to fix the issue.

Fixed that. Due to previous bug, the repository was in weird state, it
should be correct now.

-- 
	Michal Čihař | http://cihar.com | http://blog.cihar.com

**Comment by andrewshadura** [comment]
Thanks.
By the way, if you're using Mercurial 3.3.2, I think you can safely enable the following:

    [experimental]
    evolveopts=all

This will enable changeset obsoletion, so that upstream rebases will be handled better when the upstream also uses them.

**Comment by nijel** [comment]
Done in 840dc3c

**Comment by nijel** [comment]
Today, I've discovered another bug in Mercurial backend - it tend to create more heads, causing further problems. This should be fixed as of 946b80f5defd3a9b2d993ae583373ec36fe86dd5.

Also Mercurial is now covered by functional tests in d9deb7833e4b8d54b24851efdbb8894c110a4bd0 (before it had only unit tests for VCS backend).

Should there be some other issues with this backend, please report separately.

**Comment by Wohlstand** [comment]
I have found a very similar problem in the latest 3.9 version... Mercurial won't rebase local commits into the top and won't push even I'll commit to the top or when I will rebase a head manually: it always says "nothing to push". Rebasing usually says "nothing to rebase", when in fact it IS a commit to rebase which didn't rebase. I have tried to trace all output of hg and all args of hg (I modified `/usr/bin/hg` with adding capturing of `stdout` and `stderr`), and I have found that odd behavior of Mercurial in some moment it prints lots of odd changes (possibly it compares with some ancient state or other). The repository has many opened branches are created for different purposes.
